components/keys/src/types.rs

```diff
@@ -34,25 +35,26 @@ impl Key {
     /// Creates a key from raw bytes.
     #[inline]
     pub fn from_raw(key: &[u8]) -> Key {
-        // adding extra length for appending timestamp
-        let len = codec::bytes::max_encoded_bytes_size(key.len()) + codec::number::U64_SIZE;
+        let len = MemComparableByteCodec::encoded_len(key.len());
         let mut encoded = Vec::with_capacity(len);
-        encoded.encode_bytes(key, false).unwrap();
+        encoded.write_comparable_bytes(key).unwrap();
         Key(encoded)
     }
 
     /// Gets and moves the raw representation of this key.
     #[inline]
     pub fn into_raw(self) -> Result<Vec<u8>, codec::Error> {
         let mut k = self.0;
-        bytes::decode_bytes_in_place(&mut k, false)?;
+        MemComparableByteCodec::try_decode_first_in_place(&mut k)?;
         Ok(k)
     }
 
     /// Gets the raw representation of this key.
     #[inline]
     pub fn to_raw(&self) -> Result<Vec<u8>, codec::Error> {
-        bytes::decode_bytes(&mut self.0.as_slice(), false)
+        let mut k = Vec::with_capacity(self.0.len() / (MEMCMP_GROUP_SIZE + 1) * MEMCMP_GROUP_SIZE);
```

Comment:
Looks like you can simply allocate a vector with the same size as the original, to avoid the complex calculation and hide the codec implementation.

Reply:
done